tsdb/isolation.go

```diff
@@ -0,0 +1,220 @@
+	readsOpen *IsolationState
+}
+
+func newIsolation() *isolation {
+	isoState := &IsolationState{}
+	isoState.next = isoState
+	isoState.prev = isoState
+
+	return &isolation{
+		writesOpen: map[uint64]struct{}{},
+		readsOpen:  isoState,
+	}
+}
+
+// lowWatermark returns the writeId below which
+// we no longer need to track which writes were from
+// which writeId.
+func (i *isolation) lowWatermark() uint64 {
+	i.writeMtx.Lock() // Take writeMtx first.
+	defer i.writeMtx.Unlock()
+	i.readMtx.Lock()
+	defer i.readMtx.Unlock()
+	if i.readsOpen.prev == i.readsOpen {
+		return i.lastWriteID
+	}
+	return i.readsOpen.prev.lowWaterMark
+}
+
+// State returns an object used to control isolation
+// between a query and writes. Must be closed when complete.
+func (i *isolation) State() *IsolationState {
+	i.writeMtx.Lock() // Take write mutex before read mutex.
+	defer i.writeMtx.Unlock()
+	isoState := &IsolationState{
+		maxWriteID:       i.lastWriteID,
+		lowWaterMark:     i.lastWriteID,
+		incompleteWrites: make(map[uint64]struct{}, len(i.writesOpen)),
+		isolation:        i,
+	}
+	for k := range i.writesOpen {
+		isoState.incompleteWrites[k] = struct{}{}
+		if k < isoState.lowWaterMark {
+			isoState.lowWaterMark = k
+		}
+	}
+
+	i.readMtx.Lock()
+	defer i.readMtx.Unlock()
+	isoState.prev = i.readsOpen
+	isoState.next = i.readsOpen.next
+	i.readsOpen.next.prev = isoState
+	i.readsOpen.next = isoState
+	return isoState
+}
+
+// newWriteID increments the transaction counter and returns a new transaction ID.
+func (i *isolation) newWriteID() uint64 {
+	i.writeMtx.Lock()
+	i.lastWriteID++
+	writeID := i.lastWriteID
+	i.writesOpen[writeID] = struct{}{}
+	i.writeMtx.Unlock()
+
+	return writeID
+}
+
+func (i *isolation) closeWrite(writeID uint64) {
+	i.writeMtx.Lock()
+	delete(i.writesOpen, writeID)
+	i.writeMtx.Unlock()
+}
+
+// The transactionID ring buffer.
+type txRing struct {
+	txIDs     []uint64
+	txIDFirst int // Position of the first id in the ring.
+	txIDCount int // How many ids in the ring.
+}
+
+func newTxRing(cap int) *txRing {
+	return &txRing{
+		txIDs: make([]uint64, cap),
+	}
+}
+
+func (txr *txRing) add(writeID uint64) {
+	if txr.txIDCount == len(txr.txIDs) {
+		// Ring buffer is full, expand by doubling.
+		newRing := make([]uint64, txr.txIDCount*2)
+		idx := copy(newRing[:], txr.txIDs[txr.txIDFirst:])
+		copy(newRing[idx:], txr.txIDs[:txr.txIDFirst])
+		txr.txIDs = newRing
+		txr.txIDFirst = 0
+	}
+
+	txr.txIDs[(txr.txIDFirst+txr.txIDCount)%len(txr.txIDs)] = writeID
+	txr.txIDCount++
+}
+
+func (txr *txRing) cleanupWriteIDsBelow(bound uint64) {
+	pos := txr.txIDFirst
+
+	for txr.txIDCount > 0 {
+		if txr.txIDs[pos] < bound {
+			txr.txIDFirst++
+			txr.txIDCount--
+		} else {
+			break
+		}
+
+		pos++
+		if pos == len(txr.txIDs) {
+			pos = 0
+		}
+	}
+
+	txr.txIDFirst %= len(txr.txIDs)
+}
+
+// cutoffN will only keep the latest N transactions in the ring.
+// Will also downsize the ring if possible.
```

Comment:
Are we sure we want to downsize? My thinking previously was that if it got that big before, it'll get that big again.

Reply:
Again, perhaps @gouthamve has context.

Reply:
Rough thoughts from my side: If it is changing at all, that probably means there is a varying ingestion load. In steady state, it should not change that much. Which means the downsizing will help to downsize after an irregular ingestion peak, which is good. It will also downsize if ingestion stalls. Which is not desirable, but not really that harmful, either. Being able to shrink again after a spike seems to be worth the price.
(In practice, it's probably pretty irrelevant anyway.)

Reply:
Ass said above, will remove that as it doesn't seem to make a dent and appears to be buggy.